diff --git a/meson.build b/meson.build
index f380255..abd7cd3 100644
--- a/meson.build
+++ b/meson.build
@@ -39,23 +39,8 @@ add_project_arguments(
   ],
   language: 'cpp')
 
-wlroots = subproject('wlroots', default_options: ['examples=false'])
-have_xwlr = wlroots.get_variable('features').get('xwayland')
 xcb_dep = dependency('xcb', required: get_option('xwayland'))
 
-cmake = import('cmake')
-udis = cmake.subproject('udis86')
-udis86 = udis.dependency('libudis86')
-
-if get_option('xwayland').enabled() and not have_xwlr
-	error('Cannot enable Xwayland in Hyprland: wlroots has been built without Xwayland support')
-endif
-have_xwayland = xcb_dep.found() and have_xwlr
-
-if not have_xwayland
-add_project_arguments('-DNO_XWAYLAND', language: 'cpp')
-endif
-
 backtrace_dep = cpp_compiler.find_library('execinfo', required: false)
 systemd_dep = dependency('libsystemd', required: get_option('systemd'))
 
diff --git a/src/meson.build b/src/meson.build
index 7b658d3..da8baa5 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -7,7 +7,7 @@ executable('Hyprland', src,
     server_protos,
     dependency('wayland-server'),
     dependency('wayland-client'),
-    wlroots.get_variable('wlroots'),
+    dependency('wlroots'),
     dependency('cairo'),
     dependency('libdrm'),
     dependency('egl'),
@@ -16,7 +16,7 @@ executable('Hyprland', src,
     xcb_dep,
     backtrace_dep,
     systemd_dep,
-    udis86,
+    dependency('udis86'),
 
     dependency('pixman-1'),
     dependency('gl', 'opengl'),
