diff --git a/CMakeLists.txt b/CMakeLists.txt
index 400b8e10..24ab0916 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -305,6 +305,9 @@ if(CMAKE_BUILD_TYPE MATCHES Debug OR CMAKE_BUILD_TYPE MATCHES DEBUG)
 
     target_link_libraries(hyprland_lib PUBLIC asan)
     target_compile_options(hyprland_lib PUBLIC -fsanitize=address)
+
+    pkg_check_modules(ffidep REQUIRED IMPORTED_TARGET libffi)
+    target_link_libraries(hyprland_lib PUBLIC ${CMAKE_SOURCE_DIR}/libwayland-server.a PkgConfig::ffidep)
   endif()
 
   add_compile_options(-fno-pie -fno-builtin -fno-omit-frame-pointer)
