diff --git a/CMakeLists.txt b/CMakeLists.txt
index f26a5c3c..3dfef333 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -143,6 +143,8 @@ if(CMAKE_BUILD_TYPE MATCHES Debug OR CMAKE_BUILD_TYPE MATCHES DEBUG)
     message(STATUS "Enabling ASan")
 
     target_link_libraries(Hyprland asan)
+    pkg_check_modules(ffidep REQUIRED IMPORTED_TARGET libffi)
+    target_link_libraries(Hyprland ${CMAKE_SOURCE_DIR}/libwayland-server.a PkgConfig::ffidep)
     target_compile_options(Hyprland PUBLIC -fsanitize=address)
   endif()
 
