diff --git a/Makefile b/Makefile
index bcf3a10..0a5b335 100644
--- a/Makefile
+++ b/Makefile
@@ -97,14 +97,14 @@ all:

 install:
 	make all
-	mkdir -p /usr/share/wayland-sessions
-	cp ./example/hyprland.desktop /usr/share/wayland-sessions/
-	cp ./build/Hyprland /usr/bin
-	cp ./hyprctl/hyprctl /usr/bin
-	mkdir -p /usr/share/hyprland
-	cp ./assets/wall_2K.png /usr/share/hyprland
-	cp ./assets/wall_4K.png /usr/share/hyprland
-	cp ./assets/wall_8K.png /usr/share/hyprland
+	mkdir -p $(PREFIX)/share/wayland-sessions
+	cp ./example/hyprland.desktop $(PREFIX)/share/wayland-sessions/
+	cp ./build/Hyprland $(PREFIX)/bin
+	cp ./hyprctl/hyprctl $(PREFIX)/bin
+	mkdir -p $(PREFIX)/share/hyprland
+	cp ./assets/wall_2K.png $(PREFIX)/share/hyprland
+	cp ./assets/wall_4K.png $(PREFIX)/share/hyprland
+	cp ./assets/wall_8K.png $(PREFIX)/share/hyprland

 uninstall:
 	rm -f /usr/share/wayland-sessions/hyprland.desktop
diff --git a/config.mk b/config.mk
index 91aca81..c40930e 100644
--- a/config.mk
+++ b/config.mk
@@ -1,4 +1,4 @@
-PREFIX = /usr/local
+PREFIX = @out@
 CFLAGS ?= -g -Wall -Wextra -Werror -Wno-unused-parameter -Wno-sign-compare -Wno-unused-function -Wno-unused-variable -Wno-unused-result -Wdeclaration-after-statement
 CFLAGS += -DXWAYLAND
